src/commands/mod.rs

```diff
@@ -2005,14 +2008,71 @@ fn cmd_new(ui: &mut Ui, command: &CommandHelper, args: &NewArgs) -> Result<(), C
         "expected a non-empty list from clap"
     );
     let commits = resolve_base_revs(&workspace_command, &args.revisions)?;
-    let parent_ids = commits.iter().map(|c| c.id().clone()).collect();
+    let parent_ids = commits.iter().map(|c| c.id().clone()).collect_vec();
     let mut tx = workspace_command.start_transaction("new empty commit");
-    let merged_tree = merge_commit_trees(tx.base_repo().as_repo_ref(), &commits);
-    let new_commit = tx
-        .mut_repo()
-        .new_commit(command.settings(), parent_ids, merged_tree.id().clone())
-        .set_description(&args.message)
-        .write()?;
+    let mut num_rebased = 0;
+    let new_commit;
+    if args.insert_before {
+        // Instead of having the new commit as a child of the changes given on the
+        // command line, add it between the changes' parents and the changes.
+        // The parent of the new commit will be the parents of the target commits
+        // which are not descendants of other target commits.
+        let child_ids = parent_ids;
+        let children = RevsetExpression::commits(child_ids.clone());
+        let new_parents = children.parents().minus(&children.descendants());
+        let new_parents_commits = tx
+            .base_workspace_helper()
+            .evaluate_revset(&new_parents)?
+            .iter()
+            .commits(tx.base_repo().store())
+            .collect::<Result<Vec<_>, _>>()?;
+        if new_parents_commits.len() > 1
+            && new_parents_commits.contains(&tx.base_repo().store().root_commit())
+        {
+            return Err(CommandError::UserError {
+                message: "Cannot merge with root revision".to_owned(),
+                hint: None,
+            });
+        }
+        let merged_tree = merge_commit_trees(tx.base_repo().as_repo_ref(), &new_parents_commits);
+        let grandparent_ids = new_parents_commits.iter().map(|c| c.id().clone()).collect();
+        new_commit = tx
+            .mut_repo()
+            .new_commit(
+                command.settings(),
+                grandparent_ids,
+                merged_tree.id().clone(),
+            )
+            .set_description(&args.message)
+            .write()?;
+        num_rebased = child_ids.len();
+        for child_id in child_ids {
+            let parent_commit = tx
+                .base_workspace_helper()
+                .resolve_single_rev(&child_id.hex())?;
```

Comment:
You can use `tx.mut_repo().store().get_commit(&child_id)` to get the commit here. thereby avoiding the roundtrip via a revset (which is technically also incorrect because it would pick the wrong commit if the child_id is "abc123" and the repo has a branch called "abc123" pointing to "abc456").

Actually, it looks like we can even avoid the lookup from the backend here since `commits` seems to be the `parent_commit`s.

Reply:
indeed, I was surprised to have to go through `hex()` here, and it could have been incorrect.

However, I'm not sure how I could avoid the lookup, I don't see that `child_ids` have been materialized into `Commit` objects.

Reply:
If I'm reading it correctly, `child_ids` is assigned from `parents_ids`, which comes from `commits`, so doing `for parent_commit in commits` seems like it should work.

Reply:
You're reading this correctly. I've pushed a version with this change and some variable renamings to make it clearer.